hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/freon/FollowerAppendLogEntryGenerator.java

```diff
@@ -0,0 +1,470 @@
+      description = "Maximum in-flight messages",
+      defaultValue = "10")
+  private int inflightLimit;
+
+  private TimedSemaphore rateLimiter;
+
+  private RaftPeerProto requestor;
+
+  private long term = 2L;
+
+  private RaftServerProtocolServiceStub stub;
+
+  private Random callIdRandom = new Random();
+
+  private ByteString dataToWrite;
+
+  private Timer timer;
+
+  private BlockingQueue<Long> inFlightMessages;
+
+  private StreamObserver<AppendEntriesRequestProto> sender;
+
+  @Override
+  public Void call() throws Exception {
+    inFlightMessages = new LinkedBlockingQueue<>(inflightLimit);
+
+    timer = getMetrics().timer("append-entry");
+    byte[] data = RandomStringUtils.randomAscii(chunkSize)
+        .getBytes(StandardCharsets.UTF_8);
+
+    dataToWrite = ByteString.copyFrom(data);
+
+    OzoneConfiguration conf = createOzoneConfiguration();
+
+    setServerIdFromFile(conf);
+
+    Preconditions.assertTrue(getThreadNo() == 1,
+        "This test should be executed from one thread");
+
+    //the raft identifier which is used by the freon
+    requestor = RaftPeerProto.newBuilder()
+        .setId(RaftPeerId.valueOf(FAKE_LEADER_ID).toByteString())
+        .setAddress(FAKE_LEADER_ADDDRESS)
+        .build();
+
+    NettyChannelBuilder channelBuilder =
+        NettyChannelBuilder.forTarget(serverAddress);
+    channelBuilder.negotiationType(NegotiationType.PLAINTEXT);
+    ManagedChannel build = channelBuilder.build();
+    stub = RaftServerProtocolServiceGrpc.newStub(build);
+
+    if (rateLimit != 0) {
+      rateLimiter = new TimedSemaphore(1, TimeUnit.SECONDS, rateLimit);
+    }
+
+    init();
+
+    sender = stub.appendEntries(this);
+
+    if (nextIndex == 0) {
+      //first: configure a new ratis group (one follower, one fake leader
+      // (freon))
+      configureGroup();
+
+      RequestVoteReplyProto vote = requestVote().get(1000, TimeUnit.SECONDS);
+      LOG.info("Datanode answered to the vote request: {}", vote);
+      if (!vote.getServerReply().getSuccess()) {
+        throw new RuntimeException(
+            "Datanode didn't vote to the fake freon leader.");
+      }
+
+      //send the first appendEntry. This one is special as it initialized the
+      // log.
+      long callId = callIdRandom.nextLong();
+      inFlightMessages.put(callId);
+      sender.onNext(createInitialLogEntry(callId));
+
+      nextIndex = 1L;
+    }
+
+    //We can generate as mach entry as we need.
+    runTests(this::sendAppendLogEntryRequest);
+
+    if (rateLimiter != null) {
+      rateLimiter.shutdown();
+    }
+
+    return null;
+  }
+
+  /**
+   * Seend a new HB and record the call id to handle response.
+   */
+  private void sendAppendLogEntryRequest(long sequence) {
+    timer.time(() -> {
+      AppendEntriesReplyProto replyProto = null;
+      try {
+        long callId = callIdRandom.nextLong();
+        inFlightMessages.put(callId);
+        sender.onNext(createAppendLogEntry(sequence, callId));
```

Comment:
Result of `createAppendLogEntry` should be assigned to `replyProto` if it is to be used in the log below.

Reply:
Earlier I used sync call and I stored the *reply*. Result of `createAppendLogEntry` is the request and we have no *reply* here as the same observer (`sender`) is used for all the requests.

I removed the old, unused code (logging just the exception)